packages/client/src/__tests__/integration/happy/interactive-transactions-postgres/test.ts

```diff
@@ -0,0 +1,473 @@
+
+    const result = prisma.$transaction((prisma) => {
+      // we accumulate all the forbidden methods and expect to be undefined
+      return forbidden.reduce((acc, item) => acc ?? prisma[item], undefined)
+    })
+
+    await expect(result).resolves.toBe(undefined)
+  })
+
+  /**
+   * If one of the query fails, all queries should cancel
+   */
+  test('rollback query', async () => {
+    const result = prisma.$transaction(async (prisma) => {
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.create()\` invocation in
+            /client/src/__tests__/integration/happy/interactive-transactions-postgres/test.ts:0:0
+
+              184   },
+              185 })
+              186 
+            → 187 await prisma.user.create(
+              Unique constraint failed on the fields: (\`email\`)
+          `)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * A transaction that is already 'commit' cannot be reused
+   */
+  test('already committed', async () => {
+    let transactionBoundPrisma
+    await prisma.$transaction((prisma) => {
+      transactionBoundPrisma = prisma
+    })
+
+    const result = prisma.$transaction(async () => {
+      await transactionBoundPrisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`transactionBoundPrisma.user.create()\` invocation in
+            /client/src/__tests__/integration/happy/interactive-transactions-postgres/test.ts:0:0
+
+              218 })
+              219 
+              220 const result = prisma.$transaction(async () => {
+            → 221   await transactionBoundPrisma.user.create(
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Committed'.
+          `)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * Batching should work with using the interactive transaction logic
+   */
+  test('batching', async () => {
+    await prisma.$transaction([
+      prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      }),
+      prisma.user.create({
+        data: {
+          email: 'user_2@website.com',
+        },
+      }),
+    ])
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(2)
+  })
+
+  /**
+   * A bad batch should rollback using the interactive transaction logic
+   * // TODO: skipped because output differs from binary to library
```

Comment:
We may have two copies of this test, each conditionally skipped depending on `getClientEngineType()`

Reply:
Thanks for the suggestion.